hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/ListActionTaker.java

```diff
@@ -0,0 +1,178 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.fs.azurebfs.services;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.azurebfs.AbfsConfiguration;
+import org.apache.hadoop.fs.azurebfs.contracts.exceptions.AzureBlobFileSystemException;
+import org.apache.hadoop.fs.azurebfs.contracts.services.BlobListResultSchema;
+import org.apache.hadoop.fs.azurebfs.contracts.services.ListResultEntrySchema;
+import org.apache.hadoop.fs.azurebfs.contracts.services.ListResultSchema;
+import org.apache.hadoop.fs.azurebfs.utils.TracingContext;
+
+import static org.apache.hadoop.fs.azurebfs.constants.AbfsHttpConstants.ROOT_PATH;
+
+/**
+ * ListActionTaker is an abstract class that provides a way to list the paths
+ * recursively and take action on each path. The implementations of this class
+ * should provide the action to be taken on each listed path.
+ */
+public abstract class ListActionTaker {
+
+  protected final Path path;
+
+  protected final AbfsBlobClient abfsClient;
+
+  protected final TracingContext tracingContext;
+
+  private final ExecutorService executorService;
+
+  private final int maxConsumptionParallelism;
+
+  private final AtomicBoolean producerThreadToBeStopped = new AtomicBoolean(
+      false);
+
+  public ListActionTaker(Path path,
+      AbfsClient abfsClient,
+      int maxConsumptionParallelism,
+      TracingContext tracingContext) {
+    this.path = path;
+    this.abfsClient = (AbfsBlobClient) abfsClient;
+    this.tracingContext = tracingContext;
+    this.maxConsumptionParallelism = maxConsumptionParallelism;
+    executorService = Executors.newFixedThreadPool(maxConsumptionParallelism);
+  }
+
+  abstract boolean takeAction(Path path) throws IOException;
+
+  private boolean takeAction(List<Path> paths) throws IOException {
+    List<Future<Boolean>> futureList = new ArrayList<>();
+    for (Path path : paths) {
+      Future<Boolean> future = executorService.submit(() -> {
+        return takeAction(path);
+      });
+      futureList.add(future);
+    }
+
+    IOException executionException = null;
+    boolean actionResult = true;
+    for (Future<Boolean> future : futureList) {
+      try {
+        Boolean result = future.get();
+        if (!result) {
+          actionResult = false;
+        }
+      } catch (InterruptedException ignored) {
+
```

Comment:
There is nothing in the code in the future that can throw IntteruptedException. But, since, future.get() has IntteruptedException in its signature, we have to get it gracefully ignored for compiling.